[sc-30416] Allow generating predefined Slack user mapping

jakeschurch opened the pull request:
#### Description

This PR introduces robust caching and mapping of Slack user details by GitHub login, significantly reducing reliance on Slack and GitHub API calls and mitigating rate limiting issues. It incorporates extensive requested changes and feedback from review:

- **Mapping Support & Caching**
  - Add support for providing a GitHub-to-Slack user mapping via environment variable (`SLACK_DEPLOY_GITHUB_USERS`, formerly `SLACK_GITHUB_MAPPING_RAW`) or file.
  - Cache Slack user details by GitHub login, skipping API calls when mapping is available.
  - Skip adding users to cache if errors are thrown or if users are deactivated/deleted.

- **Action & Workflow Refactor**
  - Split notify and generate-mapping actions into distinct entrypoints.
  - Rename default action workflow to `notify` in dist output.
  - Introduce `.github/actions/generate-mapping/action.yml` for generating a JSON mapping of Slack users by GitHub login.
  - Refactor main entry to pass mapping input for author resolution.

- **Code Quality & Feedback**
  - Unify mapping shape validation and add type guard tests.
  - Extract Slack user lookup logic to reusable helper function.
  - Reference existing `MessageAuthor` type for mapping output.
  - Improve code readability, error handling, and caching logic.
  - Use `getEnv` for environment variable access.
  - Rename variables and inputs for clarity and consistency.
  - Remove redundant error logging and unnecessary print statements.
  - Address parallel API call concerns and clarify rate limiting handling.

- **Documentation**
  - Update README to document new environment variable, mapping usage, and benefits.
  - Add examples and clarify usage for custom mapping.

#### Testing

- [Fieldguide/infrastructure/actions/runs/17133941858/job/48605070955?pr=1045](Fieldguide/infrastructure/actions/runs/17133941858/job/48605070955?pr=1045)
- #181: Cache Slack users.list API response as artifact for deploy pipeline notifications
- #1045: HOLD: test branch
- Slack channel for testing

#### Reviewer Notes

- All requested changes from @namoscato have been addressed, including variable naming, error handling, type validation, and code extraction.
- No merge conflicts; all checks have passed.

jakeschurch committed:
feat(sc-30416): add GitHub-to-Slack user mapping generation action

NOTES:

- Introduce `.github/actions/generate-mapping/action.yml` for
  generating a JSON mapping of Slack users by GitHub login.

- Add `src/githubToSlackMapping.ts` to fetch GitHub org users,
  match them to Slack users, and output mapping.

- Refactor `src/main.ts` to support `notify` and `generate-mapping`
  modes via `__mode` input.

- Update `action.yml` to include `__mode` input for internal mode
  selection.

Signed-off-by: Jake Schurch <jakeschurch@gmail.com>

jakeschurch committed:
feat(sc-30416): add support for user mapping file to resolve Slack user by GitHub login

- Introduce `user_mapping_filepath` input in `action.yml` to optionally
  provide a mapping file for Slack user details by GitHub login.

- Update `getMessageAuthorFactory` and related logic to use the mapping
  file if provided, avoiding Slack and GitHub API calls when possible.

- Add fallback and error handling for mapping file usage.

- Refactor main entry to pass `user_mapping_filepath` to author
  resolution.

Signed-off-by: Jake Schurch <jakeschurch@gmail.com>

jakeschurch committed:
[sc-30416] exec pnpm run package

Signed-off-by: Jake Schurch <jakeschurch@gmail.com>

jakeschurch committed:
feat(sc-30416): add support for user mapping file to resolve Slack user by GitHub login

- Introduce `user_mapping_filepath` input in `action.yml` to optionally
  provide a mapping file for Slack user details by GitHub login.

- Update `getMessageAuthorFactory` and related logic to use the mapping
  file if provided, avoiding Slack and GitHub API calls when possible.

- Add fallback and error handling for mapping file usage.

- Refactor main entry to pass `user_mapping_filepath` to author
  resolution.

Signed-off-by: Jake Schurch <jakeschurch@gmail.com>

jakeschurch committed:
[sc-30416] fmt

Signed-off-by: Jake Schurch <jakeschurch@gmail.com>

jakeschurch committed:
[sc-30416] more testing

Signed-off-by: Jake Schurch <jakeschurch@gmail.com>

jakeschurch committed:
[sc-30416] add check

Signed-off-by: Jake Schurch <jakeschurch@gmail.com>

jakeschurch committed:
[sc-30416] package dist

Signed-off-by: Jake Schurch <jakeschurch@gmail.com>

jakeschurch committed:
refactor: split notify and generateMapping actions, remove mode input

- Separate notify and generateMapping logic into distinct entrypoints
  (`src/notify.ts` and new `src/generateMapping.ts`)

- Remove internal `__mode` input from action.yml files

- Update action.yml and generate-mapping/action.yml to reference new
  dist entrypoints

- Refactor package.json exports and scripts for new structure

- Improve maintainability and clarity by decoupling action
  responsibilities

refs: [sc-30416]

Signed-off-by: Jake Schurch <jakeschurch@gmail.com>

jakeschurch committed:
refactor(generate-mapping): return raw mapping JSON instead of writing to file

- Change output from file path to raw JSON string in GitHub Action
  output (`raw_mapping_json`)

- Update `generateGithubToSlackMapping` to return JSON string instead of
  writing to disk

- Remove file system operations from mapping generation logic

- Update action input/output and usage to reflect new output format

Signed-off-by: Jake Schurch <jakeschurch@gmail.com>

jakeschurch committed:
docs(readme): document SLACK_GITHUB_MAPPING_RAW for custom GitHub-to-Slack user mapping

- Add details to README about the optional SLACK_GITHUB_MAPPING_RAW
  environment variable, allowing users to provide raw JSON or YAML
  mappings of GitHub logins to Slack user details.

- Explains usage, example formats, and benefits of avoiding API calls
  for user mapping.

refs: sc-30416

Signed-off-by: Jake Schurch <jakeschurch@gmail.com>

jakeschurch committed:
[sc-30416] docs: fmt

Signed-off-by: Jake Schurch <jakeschurch@gmail.com>

jakeschurch committed:
[sc-30416] add new dist

Signed-off-by: Jake Schurch <jakeschurch@gmail.com>

jakeschurch committed:
[sc-30416] remove print statements

Signed-off-by: Jake Schurch <jakeschurch@gmail.com>

jakeschurch committed:
[sc-30416] fix for gitattrs

Signed-off-by: Jake Schurch <jakeschurch@gmail.com>

jakeschurch committed:
[sc-30416] fixup

Signed-off-by: Jake Schurch <jakeschurch@gmail.com>

namoscato commented on src/notify.ts:
```diff
@@ -12,27 +12,35 @@ async function run(): Promise<void> {
+  octokit: OctokitClient,
+  slack: SlackClient
+): Promise<void> {
+  const rawMapping = process.env.SLACK_GITHUB_MAPPING_RAW || ''
```
Thoughts on renaming this to something like `SLACK_DEPLOY_GITHUB_USERS`? I'd like to at least keep all of the environment variables prefixed with `SLACK_DEPLOY_*` for consistency.

Also minor, but could we use [`getEnv`](https://github.com/Fieldguide/action-slack-deploy-pipeline/blob/251c8b88017ee6d810f00505baef76018953a66d/src/input.ts#L22-L29)?

namoscato commented on src/getMessageAuthorFactory.ts:
```diff
@@ -5,33 +5,70 @@ import {OctokitClient} from './github/types'
+      }
+    }
+    if (mapping[githubUserLogin]) {
+      return mapping[githubUserLogin]
```
Would it be worth doing any runtime validation to assert the shape of this object is expected?

namoscato commented on src/getMessageAuthorFactory.ts:
```diff
@@ -5,33 +5,70 @@ import {OctokitClient} from './github/types'
+      try {
+        mapping = yaml.load(rawMapping) as Record<string, MessageAuthor>
+      } catch (yamlErr) {
+        warning('Failed to parse mapping as YAML.')
```
Should we let this error bubble up (and get handled below)?

namoscato commented on src/notify.ts:
```diff
@@ -12,27 +12,35 @@ async function run(): Promise<void> {
-
+    const errMsg = err instanceof Error ? err.message : String(err)
+    error(`Error: ${errMsg}`)
+    setFailed(errMsg)
```
Is this change necessary given [`error` is invoked](https://github.com/actions/toolkit/blob/f58042f9cc16bcaa87afaa86c2974a8c771ce1ea/packages/core/src/core.ts#L248) from within `setFailed`?

namoscato commented on src/getMessageAuthorFactory.ts:
```diff
@@ -5,33 +5,70 @@ import {OctokitClient} from './github/types'
+}
+
+interface GetMessageAuthorFactoryOptions {
+  rawMapping?: string
```
Thoughts on renaming this property to something like `githubUserMapping`?

namoscato commented on src/getMessageAuthorFactory.ts:
```diff
@@ -51,6 +88,19 @@ async function getMessageAuthor(
+        return messageAuthor
+      }
+      warning('No cached mapping found, fetching Slack user by GitHub name.')
+    }
```
Is there a need for `messageAuthor` outside of this block?

Given the private method already handles conditional logic, this might be a more readable alternative:

```ts
const messageAuthor = maybeGetMessageAuthorFromRawMapping(
  githubSender.login,
  rawMapping
)
if (messageAuthor) {
  return messageAuthor
}
```

namoscato commented on dist/index.js:
Should this and its sibling `licenses.txt` be removed?

namoscato commented on src/generateMapping.ts:
```diff
@@ -0,0 +1,48 @@
+  }
+}
+
+async function generateMapping(
```
Thoughts on renaming this module / action to something like `generateGitHubUserMapping`?

namoscato replied:
@jakeschurch, would you be open to making this name more specific? I'd be happy to make the change, if so.

namoscato replied:
Renamed to `generateUserMapping` in b1f2cb531594de71d10a42b004932950cdab8c31.

I also renamed `notify` → `notifySlack`.

namoscato commented on .github/actions/generate-mapping/action.yml:
```diff
@@ -0,0 +1,25 @@
+  output_path:
+    description: 'output of cache mapping'
+    required: false
+    default: 'data/github_to_slack.json'
```
```suggestion
```

namoscato commented on .github/actions/generate-mapping/action.yml:
```diff
@@ -0,0 +1,25 @@
+    description: 'Repository GITHUB_TOKEN or personal access token secret'
+    required: false
+    default: ${{ github.token }}
+  org:
```
Could we prefix this?

```suggestion
  github_org:
```

namoscato commented on .github/actions/generate-mapping/action.yml:
```diff
@@ -0,0 +1,25 @@
+    default: 'data/github_to_slack.json'
+outputs:
+  raw_mapping_json:
+    description: 'Raw JSON mapping of Slack user details by GitHub user'
```
Non-blocking, but thoughts on renaming this to `json` for simplicity (given the action's purpose is to generate a mapping)?

namoscato commented on src/generateUserMapping.ts:
```diff
@@ -0,0 +1,48 @@
+  const errorReaction = getEnv(EnvironmentVariable.SlackErrorReaction)
+  return new SlackClient({
+    token,
+    channel,
+    errorReaction
```
Should we make these `SlackClient` dependencies optional given they're unnecessary in this context?

namoscato replied:
@jakeschurch, any objections to making these `channel` / `errorReaction` dependencies optional?

namoscato replied:
Updated in 942c96ac9cc2a70b8227f0cdedc59223d99dbc36.

namoscato commented on src/githubToSlackMapping.ts:
```diff
@@ -0,0 +1,106 @@
+ * @param octokit OctokitClient instance
+ * @param slack SlackClient instance
+ * @param org GitHub organization name
+ * @param outputPath Path to output JSON file
```
```suggestion
```

namoscato commented on src/githubToSlackMapping.ts:
```diff
@@ -0,0 +1,106 @@
+ * @param org GitHub organization name
+ * @param outputPath Path to output JSON file
+ */
+export async function generateGithubToSlackMapping(
```
Minor, but could we match the module name to the exported function name?

Also, to improve readability, could we pull this public function up to the top of the module [per convention](https://www.notion.so/fieldguide/Code-Conventions-d0949e6fd7bf4cab903c254e05a619aa?source=copy_link#c2c2ecd652c64aa0902bf4b3ab428d51)?

namoscato commented on src/githubToSlackMapping.ts:
```diff
@@ -0,0 +1,106 @@
+      slack_user_id: string
+      username: string
+      icon_url: string
+    }
```
Thoughts on referencing the existing [`MessageAuthor` type](https://github.com/Fieldguide/action-slack-deploy-pipeline/blob/251c8b88017ee6d810f00505baef76018953a66d/src/slack/types.ts#L15-L22)?

namoscato commented on src/githubToSlackMapping.ts:
```diff
@@ -0,0 +1,106 @@
+              ? ghUserDetails.name.toLowerCase()
+              : undefined) && user.profile?.display_name
+        )
+    ) ?? [undefined])[0]
```
Thoughts on pulling this existing logic into a function that could be used across both contexts?

https://github.com/Fieldguide/action-slack-deploy-pipeline/blob/251c8b88017ee6d810f00505baef76018953a66d/src/getMessageAuthorFactory.ts#L115-L137

I think I'm cool with the case insensitivity change but would prefer maintaining consistency.

As a followup enhancement (outside the scope of this PR), it would be useful for this generate-mapping action to expose an optional input allowing for a manual mapping between GitHub usernames → Slack user IDs in the event names do not match.

jakeschurch replied:
extracted to function

namoscato commented on src/githubToSlackMapping.ts:
```diff
@@ -0,0 +1,106 @@
+          })
+          if (getUserResponse.status !== 200) {
+            warning(`Failed to fetch user details for ${m.login}`)
+            return null
```
Is this necessary, or will the promise reject?

namoscato commented on src/githubToSlackMapping.ts:
```diff
@@ -0,0 +1,106 @@
+        .map(async m => {
+          const getUserResponse = await octokit.rest.users.getByUsername({
+            username: m.login
+          })
```
Any concerns with [rate limiting](https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28), executing all of these in parallel?

jakeschurch replied:
@namoscato personally no, since for auth'd requests (w/ github token) [we get 5k request per hour](https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#primary-rate-limit-for-authenticated-users).

namoscato commented on src/githubToSlackMapping.ts:
```diff
@@ -0,0 +1,106 @@
+  octokit: OctokitClient,
+  slack: SlackClient,
+  org: string
+): Promise<string> {
```
Non-blocking, but would `Record` be a more appropriate return type, pulling the JSON serialization up to the caller context?

namoscato reviewed (changes requested):
Thanks for your work on this, @jakeschurch – this is looking great!

I might take a pass through the README once the dust settles.

jakeschurch committed:
refactor: unify github-to-slack mapping shape validation and add type guard tests

- Refactor mapping extraction to use `githubUserMapping` instead of
  `rawMapping`
- Validate mapping shape using new `isMessageAuthor` type guard
- Extract and use `isMemberWithProfile` and `isMessageAuthor` type
  guards across codebase
- Add tests for type guards in
  `src/slack/__tests__/types.test.ts`
- Update environment variable to `SLACK_DEPLOY_GITHUB_USERS`
- Remove redundant debug logging and simplify Slack user matching logic
- Update README

refs: [sc-30416]

Signed-off-by: Jake Schurch <jakeschurch@gmail.com>

jakeschurch committed:
[sc-30416] fixup readme

Signed-off-by: Jake Schurch <jakeschurch@gmail.com>

jakeschurch committed:
fixup|refactor(getMessageAuthorFactory): extract Slack user lookup logic to helper

- Move Slack user lookup by GitHub name to new
  `getSlackUserFromGithubName` function for reuse and clarity.
- Replace inline Slack user matching logic in `getMessageAuthorFactory`
  and `generateGithubToSlackMapping` with the new helper.
- Update `isMemberWithProfile` type guard to remove `real_name` null
  check, relying only on `display_name` and `image_48` presence.
- Minor option default value change in `getMessageAuthorFactory`
  (`githubUserMapping: undefined`).

Signed-off-by: Jake Schurch <jakeschurch@gmail.com>

jakeschurch committed:
[sc-30416] add requested feedback on action input / output names

Signed-off-by: Jake Schurch <jakeschurch@gmail.com>

jakeschurch committed:
[sc-30416] make messageAuthor caching more readable

Signed-off-by: Jake Schurch <jakeschurch@gmail.com>

jakeschurch committed:
[sc-30416] improve code readability

Signed-off-by: Jake Schurch <jakeschurch@gmail.com>

jakeschurch committed:
[sc-30416] bundle

Signed-off-by: Jake Schurch <jakeschurch@gmail.com>

jakeschurch committed:
[sc-30416] bug: skip deactivated users

Signed-off-by: Jake Schurch <jakeschurch@gmail.com>

jakeschurch committed:
[sc-30416] remove deleted

Signed-off-by: Jake Schurch <jakeschurch@gmail.com>

jakeschurch committed:
[sc-30416] skip adding users to cache if error thrown

Signed-off-by: Jake Schurch <jakeschurch@gmail.com>

jakeschurch committed:
[sc-30416] run bundle

Signed-off-by: Jake Schurch <jakeschurch@gmail.com>

namoscato commented on src/getMessageAuthorFactory.ts:
```diff
@@ -1,37 +1,129 @@
+    return mapping[githubUserLogin]
+  }
+
+  return null
```
Could we keep the `try...catch` around this? https://github.com/Fieldguide/action-slack-deploy-pipeline/pull/167#discussion_r2357146111 was referring to an unnecessary _nested_ `try...catch` in the previous state.

namoscato commented on src/getMessageAuthorFactory.ts:
```diff
@@ -1,37 +1,129 @@
+  }
+
+  if (mapping[githubUserLogin]) {
+    return mapping[githubUserLogin]
```
Should we return an invalid mapping?

namoscato committed:
Tighten type predicates

namoscato committed:
Deepen getSlackUserFromName

- move to separate module
- rename from getSlackUserFromGithubName
- filter MemberWithProfile in SlackClient

namoscato committed:
Revert setFailed changes

namoscato committed:
Simplify githubToSlackMapping

- remove unnecessary reduce
- surface empty set short circuits; move to warning
- preserve not found `null` values
- fix GitHubUser type

namoscato commented on src/slack/getSlackUserFromName.ts:
```diff
@@ -0,0 +1,31 @@
+import {info} from '@actions/core'
+import {MemberWithProfile} from './types'
+
+/**
+ * Return the Slack user with the specified GitHub `name`.
+ *
+ * @throws {Error} if a user is not found, or if multiple users have the same `name`
+ */
+export function getSlackUserFromName(
+  users: MemberWithProfile[],
+  name: string | null
+): MemberWithProfile {
```
a43deadcf18c8e8c9af8900c8e44d6f65994131f moves this shared function to a separate module.

Additionally, `SlackClient.getRealUsers()` now only returns members with profiles, simplifying caller logic.

namoscato commented on src/utils/githubToSlackMapping.ts:
```diff
@@ -0,0 +1,80 @@
+ * The users are matched by full name. If a match is not found, the value will be `null`.
+ */
+export async function githubToSlackMapping(
+  octokit: OctokitClient,
+  slack: SlackClient,
+  github_org: string
+): Promise<Record<string, MessageAuthor | null>> {
+  const githubUsers = await fetchGitHubUsers(octokit, github_org)
+  if (githubUsers.length === 0) {
+    warning('No GitHub members found.')
+    return {}
+  }
+
+  info('Fetching Slack users')
+  const slackUsers = await slack.getRealUsers()
+  if (slackUsers.length === 0) {
+    warning('No Slack users found.')
+    return {}
+  }
+
+  const result: Record<string, MessageAuthor | null> = {}
+
+  for (const githubUser of githubUsers) {
+    try {
+      const slackUser = getSlackUserFromName(slackUsers, githubUser.name)
+
+      result[githubUser.login] = {
+        slack_user_id: slackUser.id,
+        username: slackUser.profile.display_name,
+        icon_url: slackUser.profile.image_48
+      }
+    } catch (err) {
+      warning(err instanceof Error ? err.message : String(err))
+
+      result[githubUser.login] = null
+    }
+  }
+
+  return result
+}
```
4edb065d384ae1e2e057309345a195e90c99e64a simplified this implementation a bit.

Also notably, I widened the return interface to include `null` values, signifying a Slack user could not be matched. I was thinking that could enable a nice `getMessageAuthor` optimization – essentially skipping the expensive API calls when we already know a user will not match (via `null`) while still falling back to them for a presumably new / unknown user (via `undefined`).

Thoughts on that proposed behavior? I didn't update `getMessageAuthor` yet but happy to do so if you're onboard.

namoscato replied:
Updated `getMessageAuthor` in 4732a6af036898a6dc6f88f5c2162594d80cfc14.

namoscato committed:
Fix setFailed handling

namoscato reviewed (changes requested):
This is close! I pushed up a few commits and have a few more comments before I do some manual testing on my end. I'm happy to help with some of the loose ends – just let me know.

namoscato committed:
Handle null githubUserMapping values

- collapse getMessageAuthorFromGithubUserMapping into maybeGetMessageAuthorFromGithubUserMapping
- tighten type predicate
- improve error handling
- introduce GitHubUserMapping type

namoscato committed:
Favor githubUserMapping over withSlackUserId

namoscato committed:
Rename entrypoints

- notify -> notifySlack
- generateMapping -> generateUserMapping
- move non-entrypoints to utils/ namespace

namoscato committed:
Make SlackClient dependencies optional

channel and errorReaction are unnecessary in the generateUserMapping context

namoscato reviewed (commented):
Alright I think this is pretty close to code-complete. @jakeschurch, would you mind glancing at where this is landing?

I just need to do some manual testing and would like to take a pass at documentation.

namoscato committed:
Update documentation

namoscato committed:
Gracefully handle missing artifact

namoscato reviewed (approved):
Alright, I tested this out and updated some documentation. Thanks again for the work on this, @jakeschurch!

namoscato committed:
Merge branch 'main' of github.com:Fieldguide/action-slack-deploy-pipeline into jakeschurch/sc-30416/cache-slack-users-to-mitigate-api-rate-limiting

* 'main' of github.com:Fieldguide/action-slack-deploy-pipeline:
  Bump @octokit/types from 14.1.0 to 15.0.0 (#180)
  Bump @types/node from 20.12.11 to 20.19.21 (#179)
  Bump the jest group with 3 updates (#176)
  Bump typescript from 5.9.2 to 5.9.3 in the typescript group (#173)
  Bump actions/setup-node from 4 to 5 (#172)
  Bump actions/checkout from 4 to 5 (#168)
